lib/application_insights/channel/contracts/data.rb

```diff
@@ -1,48 +1,9 @@
 require_relative 'json_serializable'
 
-module ApplicationInsights
-  module Channel
-    module Contracts
-      # Data contract class for type Data.
-      class Data < JsonSerializable
-        # Initializes a new instance of the Data class.
-        def initialize(options={})
-          defaults = {
-            'baseType' => nil,
-            'baseData' => nil
-          }
-          values = {
-            'baseData' => nil
-          }
-          super defaults, values, options
-        end
-        
-        # Gets the base_type property.
-        def base_type
-          @values.fetch('baseType') { 
-            @values['baseType'] = nil
-          }
-        end
-        
-        # Sets the base_type property.
-        def base_type=(value)
-          if value == @defaults['baseType']
-            @values.delete 'baseType' if @values.key? 'baseType'
-          else
-            @values['baseType'] = value
-          end
-        end
-        
-        # Gets the base_data property.
-        def base_data
-          @values['baseData']
-        end
-        
-        # Sets the base_data property.
-        def base_data=(value)
-          @values['baseData'] = value
-        end
-      end
-    end
+module ApplicationInsights::Channel::Contracts
+  class Data
+    include JsonSerializable
+
+    attributes :baseType, :baseData
```

Comment:
If you discard the **Example 3** in my opinion the most readable, descriptive and simple syntax is **Example 2 B**

The problem with your proposal is that:
- you must send a mapping hash in the `initialize` method
- you have to define the accessor **only** for those who has no default value
- you have to define the _setters_ for the attributes with default values which are defined later.

It is a little complex. But the big issue here is that the object inherits from `JsonSerializable` which goes against the OOP practices.